client/allocrunner/taskrunner/envoy_version_hook.go

```diff
@@ -0,0 +1,170 @@
+package taskrunner
+
+import (
+	"context"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-version"
+	ifs "github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/consul"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/pkg/errors"
+)
+
+const (
+	// envoyVersionHookName is the name of this hook and appears in logs.
+	envoyVersionHookName = "envoy_version"
+
+	// envoyLegacyImage is used when the version of Consul is too old to support
+	// the SupportedProxies field in the self API.
+	//
+	// This is the version defaulted by Nomad before v0.13.0 and/or when using versions
+	// of Consul before v1.7.8, v1.8.5, and v1.9.0.
+	envoyLegacyImage = "envoyproxy/envoy:v1.11.2@sha256:a7769160c9c1a55bb8d07a3b71ce5d64f72b1f665f10d81aa1581bc3cf850d09"
+)
+
+type envoyVersionHookConfig struct {
+	alloc         *structs.Allocation
+	proxiesClient consul.SupportedProxiesAPI
+	logger        hclog.Logger
+}
+
+func newEnvoyVersionHookConfig(alloc *structs.Allocation, proxiesClient consul.SupportedProxiesAPI, logger hclog.Logger) *envoyVersionHookConfig {
+	return &envoyVersionHookConfig{
+		alloc:         alloc,
+		logger:        logger,
+		proxiesClient: proxiesClient,
+	}
+}
+
+// envoyVersionHook is used to determine and set the Docker image used for Consul
+// Connect sidecar proxy tasks. It will query Consul for a set of preferred Envoy
+// versions if the task image is unset or references ${NOMAD_envoy_version}. Nomad
+// will fallback the image to the previous default Envoy v1.11.2 if Consul is too old
+// to support the supported proxies API.
+type envoyVersionHook struct {
+	// alloc is the allocation with the envoy task being rewritten.
+	alloc *structs.Allocation
+
+	// proxiesClient is the subset of the Consul API for getting information
+	// from Consul about the versions of Envoy it supports.
+	proxiesClient consul.SupportedProxiesAPI
+
+	// logger is used to log things.
+	logger hclog.Logger
+}
+
+func newEnvoyVersionHook(c *envoyVersionHookConfig) *envoyVersionHook {
+	return &envoyVersionHook{
+		alloc:         c.alloc,
+		proxiesClient: c.proxiesClient,
+		logger:        c.logger.Named(envoyVersionHookName),
+	}
+}
+
+func (envoyVersionHook) Name() string {
+	return envoyVersionHookName
+}
+
+func (h *envoyVersionHook) Prestart(_ context.Context, request *ifs.TaskPrestartRequest, response *ifs.TaskPrestartResponse) error {
+	if h.skip(request) {
+		response.Done = true
+		return nil
+	}
+
+	// We either need to acquire Consul's preferred Envoy version or fallback
+	// to the legacy default. Query Consul and use the (possibly empty) result.
+	proxies, err := h.proxiesClient.Proxies()
```

Comment:
If Consul is being restarted when this is called it will fail the task and go into a restart loop which... seems ok?

In configurations where Consul does not start before Nomad there could be spurious task failures, but I think that's already true for Connect? Bootstrap has that awkward bulitin retry that avoids spurious task failures, but it's not a very robust approach.

Reply:
The envoy bootstrap hook's retry block I believe really targets the race condition between launching the sidecar and Nomad actually registering the service to be proxied - potentially a common occurrence. I'm thinking `/v1/agent/self` response failures are either rare enough or catastrophic enough that letting the task do a restart in those cases is probably fine